rpc/server.go

```diff
@@ -51,25 +61,51 @@ type Server struct {
 	addr           net.Addr              // Server address; may change if picking unused port
 	closed         bool                  // Set upon invocation of Close()
 	closeCallbacks []func(conn net.Conn) // Slice of callbacks to invoke on conn close
+	methods        map[string]method
 }
 
 // NewServer creates a new instance of Server.
 func NewServer(addr net.Addr, context *Context) *Server {
 	s := &Server{
-		Server:  rpc.NewServer(),
 		context: context,
 		addr:    addr,
+		methods: map[string]method{},
 	}
 	heartbeat := &HeartbeatService{
 		clock:              context.localClock,
 		remoteClockMonitor: context.RemoteClocks,
 	}
-	if err := s.RegisterName("Heartbeat", heartbeat); err != nil {
+	if err := heartbeat.Register(s); err != nil {
 		log.Fatalf("unable to register heartbeat service with RPC server: %s", err)
 	}
 	return s
 }
 
+// Register a new method handler. `name` is a qualified name of the form "Service.Name".
+// `handler` is a function that takes an argument of the same type as `reqPrototype`.
+// Both the argument and return value of 'handler' should be a pointer to a protocol
+// message type.
+func (s *Server) Register(name string, handler func(interface{}) (interface{}, error),
+	reqPrototype interface{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.methods[name]; ok {
+		return util.Errorf("method %s already registered", name)
+	}
+	reqType := reflect.TypeOf(reqPrototype)
+	if reqType.Kind() != reflect.Ptr {
```

Comment:
In the current version of gogoproto, all implementers of `gogoproto.Message` are pointer types, but this is not necessarily guaranteed. We don't currently support hypothetical value types that implement the Message interface because we'd need to change the reflection logic in `readRequest` to do so. This is now an extra sanity check to guard against any introduction of value types that implement `gogoproto.Message`. 

It is theoretically possible for a value type to implement `gogoproto.Message` (although this doesn't happen in the current version of gogoproto). If this were to happen we would need to change the reflection code in `readRequest`.